core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -519,7 +519,14 @@ public Settings archiveUnknownOrInvalidSettings(
             try {
                 Setting<?> setting = get(entry.getKey());
                 if (setting != null) {
-                    setting.get(settings);
+                    try {
+                        setting.get(settings);
+                    } catch (IllegalArgumentException ex) {
+                        if (setting.isMandatory()) {
+                            throw new IllegalStateException("can't archive mandatory setting [" + setting.getKey() + "]", ex);
```

Comment:
Also- I traced the code and as far as I can tell, in the case of `index.number_of_shards` we never say which index was problematic? I think it's important to add that info..